art/attacks/extraction/copycat_cnn.py

```diff
@@ -0,0 +1,151 @@
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements the copycat cnn attack `CopycatCNN`.
+
+| Paper link: https://arxiv.org/abs/1806.05476
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+
+import numpy as np
+
+from art import NUMPY_DTYPE
+from art.attacks.attack import Attack
+from art.classifiers.classifier import Classifier
+from art.utils import to_categorical
+
+
+logger = logging.getLogger(__name__)
+
+
+class CopycatCNN(Attack):
+    """
+    Implementation of the copycat cnn attack from Jacson et al. (2018).
+
+    | Paper link: https://arxiv.org/abs/1806.05476
+    """
+    attack_params = Attack.attack_params + ['batch_size', 'nb_epochs', 'nb_stolen']
+
+    def __init__(self, classifier, batch_size=1, nb_epochs=10, nb_stolen=1):
+        """
+        Create a copycat cnn attack instance.
+
+        :param classifier: A victim classifier.
+        :type classifier: :class:`.Classifier`
+        :param batch_size: Size of batches.
+        :type batch_size: `int`
+        :param nb_epochs: Number of epochs to use for training.
+        :type nb_epochs: `int`
+        :param nb_stolen: Number of examples to be stolen.
+        :type nb_stolen: `int`
+        """
+        super(CopycatCNN, self).__init__(classifier=classifier)
+
+        params = {'batch_size': batch_size,
+                  'nb_epochs': nb_epochs,
+                  'nb_stolen': nb_stolen}
+        self.set_params(**params)
+
+    def generate(self, x, y=None, **kwargs):
+        """
+        Generate a thieved classifier.
+
+        :param x: An array with the source input to the victim classifier.
+        :type x: `np.ndarray`
+        :param y: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes) or indices of shape
+                  (nb_samples,). Not used in this attack.
+        :type y: `np.ndarray` or `None`
+        :param thieved_classifier: A thieved classifier to be stolen.
+        :type thieved_classifier: :class:`.Classifier`
+        :return: The stolen classifier.
+        :rtype: :class:`.Classifier`
```

Comment:
`generate` as defined in the abstract base class `Attack` defines the return type as an `np.ndarray` (of adversarial examples). Maybe we should define a new method that returns a classifier?

Reply:
@beat-buesser I agree, should we define an abstract method in the Attack class, so that will create a standard for all model extraction attacks, I suggest the method name "extract"?

Reply:
Solved after discussion with @minhitbk with #250

Reply:
This is done.